tools/testsys/src/wait.rs

```diff
@@ -0,0 +1,352 @@
+            .context(error::TestsysClientSnafu)?
+        {
+            if test.agent_status().task_state != TaskState::Error
+                && (test.agent_status().results.is_empty()
+                    || test
+                        .agent_status()
+                        .results
+                        .iter()
+                        .any(|res| res.outcome == Outcome::Pass))
+            {
+                info!("The test '{test_name}' was restarted. Continuing ...")
+            }
+        }
+
+        info!(
+            "Still waiting for test '{test_name}' to be corrected. Sleeping {WAIT_CHECK_SECONDS}s"
+        );
+        tokio::time::sleep(Duration::from_secs(WAIT_CHECK_SECONDS)).await;
+    }
+}
+
+/// Wait until the resource has completed creation.
+async fn wait_for_resource_creation(
+    client: &TestManager,
+    resource_name: &str,
+) -> Result<TestRunResult> {
+    loop {
+        let resource = fetch_resources(client, &[resource_name.to_string()])
+            .await?
+            .pop()
+            .context(error::MissingSnafu {
+                item: resource_name,
+                what: "TestSys cluster",
+            })?;
+
+        if let Some(e) = resource.creation_error() {
+            error!(
+                "Resource '{}' had a creation error '{}'.",
+                resource.object_name(),
+                e
+            );
+            return Ok(TestRunResult::ResourceCreationError {
+                resource: resource_name.to_string(),
+                error: e.to_owned(),
+            });
+        }
+
+        if matches!(resource.creation_task_state(), TaskState::Completed) {
+            return Ok(TestRunResult::Successful {
+                results: Default::default(),
+            });
+        }
+
+        info!("Still waiting for resource '{resource_name}' to be created. Sleeping {WAIT_CHECK_SECONDS}s");
+        tokio::time::sleep(Duration::from_secs(WAIT_CHECK_SECONDS)).await;
+    }
+}
+
+/// Try to fetch the requested resources with retries. Missing resources will be ignored.
+async fn fetch_resources(client: &TestManager, resource_names: &[String]) -> Result<Vec<Resource>> {
+    let resource_client = client.resource_client();
+    let mut count = 0;
+    let n_iters = 5;
+    loop {
+        let resources = join_all(
+            resource_names
+                .iter()
+                .map(|resource_name| async { resource_client.get(resource_name.clone()).await }),
+        )
+        .await
+        .into_iter()
+        .filter_map(|resource_result| {
+            resource_result
+                .allow_not_found(|_| warn!("A resource could not be found."))
+                .context(error::TestsysClientSnafu)
+                .transpose()
+        })
+        .collect::<Result<Vec<Resource>>>();
+
+        match resources {
+            resources @ Ok(_) => break resources,
+            Err(_) if count < n_iters => {
```

Comment:
Similar feedback here to the other fetch loops. We probably should be using retries with backoff and jitter. I also think it would be wise to funnel our client calls through rate-limiting algorithm (I like [governor](https://crates.io/crates/governor)).